### PR TITLE
Implement Response.IsError

### PR DESCRIPTION
**Pull request by annelo-msft**
As part of https://github.com/Azure/azure-sdk-for-net/issues/23372#issuecomment-920435015, we'll add an `IsError` property to `Response`.  This is currently being implemented as an extension method in Experimental as a proof of concept before moving it into Azure.Core.

`IsError` must be set before `Request` is disposed because it takes the `HttpMessage`.  We also want to make the value of `IsError` available on `Response` in the pipeline, so we call `EvaluateError()` in a pipeline policy just after the response is received by the transport.

Per @pakrym, client request id can be set on response in the same pipeline policy.

### CODE REVIEW

File: sdk/core/Azure.Core/src/Response.cs [hunk 1]
```diff
@@ -111,6 +111,16 @@ public virtual BinaryData Content
         /// <returns>The <see cref="IEnumerable{T}"/> enumerating <see cref="HttpHeader"/> in the response.</returns>
         protected internal abstract IEnumerable<HttpHeader> EnumerateHeaders();
 
+        /// <summary>
+        /// Indicates whether this response is an error, according to the REST API.
+        /// </summary>
+        public bool IsError { get; protected set; }
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Response.cs:117]: private set?

File: sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs [hunk 2]
```diff
@@ -70,7 +70,9 @@ public ValueTask SendAsync(HttpMessage message, CancellationToken cancellationTo
         {
             message.CancellationToken = cancellationToken;
             AddHttpMessageProperties(message);
-            return _pipeline.Span[0].ProcessAsync(message, _pipeline.Slice(1));
+            var value = _pipeline.Span[0].ProcessAsync(message, _pipeline.Slice(1));
+            message.Response.EvaluateError(message);
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs:74]: One unfortunate downside of this model is that policies won't see a correct IsError value.

Comment (annelo-msft) [reply sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs:74]: Would it work to put it in the transport policy and evaluate it when the Response is created?  I guess that would mean any user-specified transport would have to know to do this.

File: sdk/core/Azure.Core.Experimental/src/ClassifiedResponse.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        private bool _disposed;
+
+        private Response Response { get; }
+
+        /// <summary>
+        /// </summary>
+        public bool IsError { get; private set; }
+
+        internal void EvaluateError(HttpMessage message)
```
Comment (pakrym) [inline sdk/core/Azure.Core.Experimental/src/ClassifiedResponse.cs:26]: Do this in ClassifiedResponse  ctor.

Comment (pakrym) [reply sdk/core/Azure.Core.Experimental/src/ClassifiedResponse.cs:26]: Ah, it takes a message.

File: sdk/core/Azure.Core.Experimental/src/ResponseExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    /// Extensions for experimenting with Response API.
+    /// </summary>
+    public static class ResponseExtensions
+    {
+        /// <summary>
+        /// This will be a property on the non-experimental Azure.Core.Response.
+        /// </summary>
+        /// <param name="response"></param>
+        /// <returns></returns>
+        public static bool IsError(this Response response) => ((ClassifiedResponse)response).IsError;
```
Comment (pakrym) [inline sdk/core/Azure.Core.Experimental/src/ResponseExtensions.cs:18]: Do a soft `as` cast so it doesn't throw when Response is not a ClassifiedResponse

File: sdk/core/Azure.Core.Experimental/tests/PipelineTests.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+            public override bool IsErrorResponse(HttpMessage message)
+            {
+                return IsRetriableResponse(message);
+            }
+        }
+
+        #region Helpers
+
+        private void SerializePet(ref Utf8JsonWriter writer, Pet pet)
```
Comment (pakrym) [inline sdk/core/Azure.Core.Experimental/tests/PipelineTests.cs:93]: rem

File: sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs (hunk 2 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs]: Updated with a policy added to the pipeline that calls Response.EvaluateError()

**Review by pakrym — APPROVED** [review]

File: sdk/core/Azure.Core.Experimental/src/ClassifiedResponse.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core.Experimental/src/ClassifiedResponse.cs:26]: Should we add overloads to the classifier that work on Response (as opposed to on messages)?

File: sdk/core/Azure.Core.Experimental/src/ResponseExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        /// </summary>
+        /// <param name="response"></param>
+        /// <returns></returns>
+        public static bool IsError(this Response response)
+        {
+            var classifiedResponse = response as ClassifiedResponse;
+
+            if (classifiedResponse == null)
+            {
+                throw new InvalidOperationException("IsError was not set on the response. " +
```
[the remark this answers is not in the document]
Comment (annelo-msft) [reply sdk/core/Azure.Core.Experimental/src/ResponseExtensions.cs:26]: When it moves to Azure.Core, the IsError property will be a direct member of response, and the intention is to set it for all responses as soon as possible after the response is received by the transport.

File: sdk/core/Azure.Core.Experimental/src/ResponsePropertiesPolicy.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+using System;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// </summary>
+    internal class ResponsePropertiesPolicy : HttpPipelinePolicy
```
[the remark this answers is not in the document]
Comment (annelo-msft) [reply sdk/core/Azure.Core.Experimental/src/ResponsePropertiesPolicy.cs:12]: Let's discuss this topic in this issue in order to unblock the PR: https://github.com/Azure/azure-sdk-for-net/issues/24164

**Pull request merged**